firestore-counter/clients/dart/distributed_counter.dart

```diff
@@ -0,0 +1,82 @@
+import 'package:async/async.dart';
+import 'package:cloud_firestore/cloud_firestore.dart';
+
+const SHARD_COLLECTION_ID = '_counter_shards_';
+
+class DistributedCounter {
+  /// Initialize the shardId, which is shared among all the instances.
+  static void init(String shardId) {
+    DistributedCounter.shardId = shardId;
+  }
+
+  /// Constructs a sharded counter object that references to a field
+  /// in a document that is a counter.
+  ///
+  /// @param doc A reference to a document with a counter field.
+  /// @param field A path to a counter field in the above document.
+  factory DistributedCounter(DocumentReference<Object?> doc, String field) {
+    final shardsRef = doc.collection(SHARD_COLLECTION_ID);
+    Map<String, int> shards = {};
+    shards[doc.path] = 0;
+    shards[shardsRef.doc(shardId).path] = 0;
+    shards[shardsRef.doc('\t' + shardId.substring(0, 4)).path] = 0;
+    shards[shardsRef.doc('\t\t' + shardId.substring(0, 3)).path] = 0;
+    shards[shardsRef.doc('\t\t\t' + shardId.substring(0, 2)).path] = 0;
+    shards[shardsRef.doc('\t\t\t\t' + shardId.substring(0, 1)).path] = 0;
+    Stream<int> snapshots = StreamGroup.mergeBroadcast<int>(shards.keys.map(
+        (path) => doc.firestore
+                .doc(path)
+                .snapshots()
+                .map<int>((DocumentSnapshot snap) {
+              shards[snap.reference.path] = snap.exists ? snap.get(field) : 0;
+              return shards.values.reduce((a, b) => a + b);
+            })));
+    return DistributedCounter._(doc, field, snapshots);
+  }
+
+  DistributedCounter._(this.doc, this.field, this._snapshots);
```

Comment:
```suggestion
  DistributedCounter(
    this.doc,
    this.field,
  ) {
    this.shardId = Uuid().v4();
    final shardsRef = doc.collection(SHARD_COLLECTION_ID);
    shards[doc.path] = 0;
    shards[shardsRef.doc(shardId).path] = 0;
    shards[shardsRef.doc('\t' + shardId.substring(0, 4)).path] = 0;
    shards[shardsRef.doc('\t\t' + shardId.substring(0, 3)).path] = 0;
    shards[shardsRef.doc('\t\t\t' + shardId.substring(0, 2)).path] = 0;
    shards[shardsRef.doc('\t\t\t\t' + shardId.substring(0, 1)).path] = 0;
  }
```

factory might be a better choice here, but this is a quicker solution.